### ISSUE TITLE
client disconnect as soon as connected with server

### ISSUE THREAD

**Issue by zychongqing** [robbiehanson/CocoaAsyncSocket#391]
After the app initial, the "didConnectToHost" already be called by system, that means it already connected with server. but immerdiately the client will disconnct this connection. After tracing, I found the function "CFReadStreamCallback" and "CFWriteStreamCallback" is called as soon as the connection built, but the type is "kCFStreamEventErrorOccurred", and the error information is "The operation couldn’t be completed. Socket is not connected".

Actually this time the client don’t have any write and read operation, and the server also don’t send any message on this new built connection. 

And if I ignore the "kCFStreamEventErrorOccurred" message, and shield “[asyncSocket closeWithError:error];" on function "CFReadStreamCallback" and "CFWriteStreamCallback", everything work well. the message between client and server can communicated normal.

**Comment by ggb88** [comment]
I have encountered the same issue - and being a newbie to this library, I've had to trace through it as well.  I've found the cause of the offending error condition in my case but not quite sure how to solve it (yet).  After the socket is established, the first thing I request is to begin reading from the socket.  This in turn is calling the below function.  Its the last "else" that generates the issue (and which I've subsequently change for the time being)



    - (void)doReadEOF
    {
	LogTrace();
	// This method may be called more than once.
	// If the EOF is read while there is still data in the preBuffer,
	// then this method may be called continually after invocations of doReadData to see if it's time to disconnect.
	
	flags |= kSocketHasReadEOF;
	
	if (flags & kSocketSecure)
	{
		// If the SSL layer has any buffered data, flush it into the preBuffer now.
		
		[self flushSSLBuffers];
	}
	
	BOOL shouldDisconnect = NO;
	NSError *error = nil;
	
	if ((flags & kStartingReadTLS) || (flags & kStartingWriteTLS))
	{
		// We received an EOF during or prior to startTLS.
		// The SSL/TLS handshake is now impossible, so this is an unrecoverable situation.
		
		shouldDisconnect = YES;
		
		if ([self usingSecureTransportForTLS])
		{
			error = [self sslError:errSSLClosedAbort];
		}
	}
	else if (flags & kReadStreamClosed)
	{
		// The preBuffer has already been drained.
		// The config allows half-duplex connections.
		// We've previously checked the socket, and it appeared writeable.
		// So we marked the read stream as closed and notified the delegate.
		// 
		// As per the half-duplex contract, the socket will be closed when a write fails,
		// or when the socket is manually closed.
		
		shouldDisconnect = NO;
	}
	else if ([preBuffer availableBytes] > 0)
	{
		LogVerbose(@"Socket reached EOF, but there is still data available in prebuffer");
		
		// Although we won't be able to read any more data from the socket,
		// there is existing data that has been prebuffered that we can read.
		
		shouldDisconnect = NO;
	}
	else if (config & kAllowHalfDuplexConnection)
	{
		// We just received an EOF (end of file) from the socket's read stream.
		// This means the remote end of the socket (the peer we're connected to)
		// has explicitly stated that it will not be sending us any more data.
		// 
		// Query the socket to see if it is still writeable. (Perhaps the peer will continue reading data from us)
		
		int socketFD = (socket4FD == SOCKET_NULL) ? socket6FD : socket4FD;
		
		struct pollfd pfd[1];
		pfd[0].fd = socketFD;
		pfd[0].events = POLLOUT;
		pfd[0].revents = 0;
		
		poll(pfd, 1, 0);
		
		if (pfd[0].revents & POLLOUT)
		{
			// Socket appears to still be writeable
			
			shouldDisconnect = NO;
			flags |= kReadStreamClosed;
			
			// Notify the delegate that we're going half-duplex
			
			__strong id theDelegate = delegate;

			if (delegateQueue && [theDelegate respondsToSelector:@selector(socketDidCloseReadStream:)])
			{
				dispatch_async(delegateQueue, ^{ @autoreleasepool {
					
					[theDelegate socketDidCloseReadStream:self];
				}});
			}
		}
		else
		{
			shouldDisconnect = YES;
		}
	}
	else
	{
        //ggb - changed to NO temporarily to avoid error condition
		shouldDisconnect = NO;
	}

**Comment by ggb88** [comment]
Well in my circumstances, it was a simple fix (after I found the issue).  I wasn't retaining the socket in a strong iVar.  Problem solved.

**Comment by zychongqing** [comment]
I think we are different issue, because I aready set this socket as strong. I tried 2 days, but still cannot find the reason, then chose a ugly way to ignore it. I changed the last part of “CFReadStreamCallback” and “CFWriteStreamCallback” function in GCDAsyncSocket.m as below:
if ((asyncSocket->flags & kStartingReadTLS) && (asyncSocket->flags & kStartingWriteTLS))
{
    [asyncSocket cf_abortSSLHandshake:error];
}
else
{
    if ((type == kCFStreamEventErrorOccurred) &&
        ([[error localizedFailureReason] isEqualToString:@"Socket is not connected"]))
    {
        if (![asyncSocket isConnected])
        {
            [asyncSocket closeWithError:error];
        }
    }
}

until now, it works well. But I still want find the real reason. Who can help me.
